h5pt.go

```diff
@@ -102,147 +84,27 @@ func (t *Table) ReadPackets(start, nrecords int, data interface{}) error {
-			if err := extractValues(ad, sfv); err != nil {
-				return err
-			}
-			// Reset the offset to the correct array sized
-			ad.offset = offset + C.size_t(rt.Size())
-		}
-		ad.numRecords = 1
-		return nil
-
-	case reflect.String:
-		ad.numRecords = 1
-		stringData := C.CString(rv.String())
-		ad.strs = append(ad.strs, unsafe.Pointer(stringData))
-		dataPtr = unsafe.Pointer(&stringData)
-		dataSize = C.size_t(unsafe.Sizeof(dataPtr))
-
-	case reflect.Ptr:
-		ad.numRecords = 1
-		return extractValues(ad, rv.Elem())
-
-	case reflect.Int8:
-		ad.numRecords = 1
-		val := C.int8_t(rv.Int())
-		dataPtr = unsafe.Pointer(&val)
-		dataSize = 1
-
-	case reflect.Uint8:
-		ad.numRecords = 1
-		val := C.uint8_t(rv.Uint())
-		dataPtr = unsafe.Pointer(&val)
-		dataSize = 1
-
-	case reflect.Int16:
-		ad.numRecords = 1
-		val := C.int16_t(rv.Int())
-		dataPtr = unsafe.Pointer(&val)
-		dataSize = 2
-
-	case reflect.Uint16:
-		ad.numRecords = 1
-		val := C.uint16_t(rv.Uint())
-		dataPtr = unsafe.Pointer(&val)
-		dataSize = 2
-
-	case reflect.Int32:
-		ad.numRecords = 1
-		val := C.int32_t(rv.Int())
-		dataPtr = unsafe.Pointer(&val)
-		dataSize = 4
-
-	case reflect.Uint32:
-		ad.numRecords = 1
-		val := C.uint32_t(rv.Uint())
-		dataPtr = unsafe.Pointer(&val)
-		dataSize = 4
-
-	case reflect.Int64:
-		ad.numRecords = 1
-		val := C.int64_t(rv.Int())
-		dataPtr = unsafe.Pointer(&val)
-		dataSize = 8
-
-	case reflect.Uint64:
-		ad.numRecords = 1
-		val := C.uint64_t(rv.Uint())
-		dataPtr = unsafe.Pointer(&val)
-		dataSize = 8
-
-	case reflect.Float32:
-		ad.numRecords = 1
-		val := C.float(rv.Float())
-		dataPtr = unsafe.Pointer(&val)
-		dataSize = 4
-
-	case reflect.Float64:
-		ad.numRecords = 1
-		val := C.double(rv.Float())
-		dataPtr = unsafe.Pointer(&val)
-		dataSize = 8
+// Struct values must only have exported fields, otherwise Append will panic.
+func (t *Table) Append(args ...interface{}) error {
+	if len(args) == 0 {
+		return fmt.Errorf("hdf5: no arguments passed to packet table append.")
+	}
 
-	case reflect.Bool:
-		ad.numRecords = 1
-		val := C.uchar(0)
-		if rv.Bool() {
-			val = 1
+	var enc cmem.Encoder
+	for _, arg := range args {
+		if err := enc.Encode(arg); err != nil {
+			return err
 		}
-		dataPtr = unsafe.Pointer(&val)
-		dataSize = 1
-
-	default:
-		return fmt.Errorf("hdf5: PT Append does not support datatype (%s).", rt.Kind())
 	}
 
-	ad.memSize = dataSize + ad.offset
-	ad.ptr = C.realloc(ad.ptr, ad.memSize)
-	C.memcpy(unsafe.Pointer(uintptr(ad.ptr)+uintptr(ad.offset)), dataPtr, dataSize)
-	ad.offset += dataSize
-
-	return nil
-}
-
-// Append appends packets to the end of a packet table.
-//
-// Struct values must only have exported fields, otherwise Append will panic.
-func (t *Table) Append(data interface{}) error {
-	var ad appendData
-	defer ad.free()
-
-	if err := extractValues(&ad, data); err != nil {
-		return err
+	if len(enc.Buf) < 0 {
+		return fmt.Errorf("hdf5: the encoded buffer is empty and can " +
```

Comment:
I feel like the error message is bit longwinded.
I would go with `"hdf5: invalid empty buffer"`

but I am not strongly opiniated about this :)